prow/cmd/hmac/main.go

```diff
@@ -0,0 +1,335 @@
+
+	currentHmacMap map[string]github.HmacsForRepo
+	newHmacConfig  config.ManagedWebhooks
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	kc, err := o.kubernetes.InfrastructureClusterClient(o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating Kubernetes client for infrastructure cluster.")
+	}
+
+	currentHmacYaml, err := getCurrentHmacTokens(kc, o.namespace, o.hmacTokenSecretName, o.hmacTokenKey)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting the current hmac yaml.")
+	}
+
+	currentHmacMap := map[string]github.HmacsForRepo{}
+	if err := yaml.Unmarshal(currentHmacYaml, &currentHmacMap); err != nil {
+		logrus.WithError(err).Fatal("Couldn't unmarshal the hmac secret as hierarchical file.")
+	}
+	configAgent := &config.Agent{}
+	if err := configAgent.Start(o.configPath, ""); err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+	newHmacConfig := configAgent.Config().ManagedWebhooks
+
+	c := &client{
+		kubernetesClient: kc,
+		options:          o,
+
+		currentHmacMap: currentHmacMap,
+		newHmacConfig:  newHmacConfig,
+	}
+
+	if err := c.handleConfigUpdate(); err != nil {
+		logrus.WithError(err).Fatal("Error handling hmac config update.")
+	}
+}
+
+func (c *client) handleConfigUpdate() error {
+	repoAdded := map[string]config.ManagedWebhookInfo{}
+	repoRemoved := map[string]bool{}
+	repoRotated := map[string]config.ManagedWebhookInfo{}
+
+	for repoName, hmacConfig := range c.newHmacConfig {
+		if _, ok := c.currentHmacMap[repoName]; ok {
+			repoRotated[repoName] = hmacConfig
+		} else {
+			repoAdded[repoName] = hmacConfig
+		}
+	}
+
+	for repoName := range c.currentHmacMap {
+		if _, ok := c.newHmacConfig[repoName]; !ok {
+			repoRemoved[repoName] = true
+		}
+	}
+
+	if err := c.handleRemovedRepo(repoRemoved); err != nil {
+		return fmt.Errorf("error handling hmac update for removed repos: %v", err)
+	}
+	if err := c.handleAddedRepo(repoAdded); err != nil {
+		return fmt.Errorf("error handling hmac update for new repos: %v", err)
+	}
+	if err := c.handledRotatedRepo(repoRotated); err != nil {
+		return fmt.Errorf("error handling hmac rotations for the repos: %v", err)
+	}
+
+	// Update the secret
+	if err := c.updateHmacTokens(); err != nil {
+		return fmt.Errorf("error updating hmac tokens: %v", err)
+	}
+
+	return nil
+}
+
+// handleRemoveRepo handles webhook removal and hmac token removal from k8s cluster for all repos removed from the declarative config.
+func (c *client) handleRemovedRepo(removed map[string]bool) error {
+	repos := make([]string, len(removed))
+	i := 0
+	for k := range removed {
+		repos[i] = k
+		i++
+	}
+
+	o := &updater.Options{
+		GitHubOptions: c.options.github,
+		Repos:         prowflagutil.NewStrings(repos...),
+		ShouldDelete:  true,
+		Confirm:       true,
+	}
```

Comment:
I have add the hook URL here, but I'm not sure how to unit test it. Do we have a mock github client that is able to handle fake webhooks?

Reply:
I have added unit tests for the `handleRemovedRepo` function. 
I'm leaving the unit tests for other functions as `TODO` since this PR is already pretty large..

Reply:
It isn't acceptable to make the fake treat an invalid deletion as a no-op unless that is how the GH API actually behaves (which I doubt is the case). If you want to avoid erroring when deleting a webhook that has already been deleted then that needs to be handled by the hmac component, not the GH API fake we are using in testing. Logging a warning and continuing if we try to delete a webhook that doesn't exist is probably sufficient.

I don't think we should leave unit tests as TODOs when adding new code, especially something like this that has the potential to take down Prow.

Reply:
Totally make sense. I have changed the fake deletion to throw out an error if the hook does not exist, also added unit tests for to cover all the key functions in this tool.